jembi/JeMPI
# Manual tallying of m and u via linker

**Ticket:** JMPI-754
**Other Related Tickets:** N/A
**Other Related PRS:**  https://github.com/jembi/JeMPI/pull/162

----

**Please Note: The PR required https://github.com/jembi/JeMPI/pull/162 to be merged first**

## Describe of changes

This change introduces the manual tallying up of m and u via linker. This happens when the linker tries to get the best canidate for an interaction. As a part of this process, it then keep a tally of Field-Equal-Pair-Match matrix (which is then used to calculate the m and u at runtime. 

## How to test / configure

The simplest way to thest this is to use the AsyncReciever, by dropping a interaction csv into the csv folder (usually under `devops/linux/docker/docker_data/data-apps/async_receiver/csv` in the dev enviroment). The async reciever will process the file, passing interation to the controller, which will then pass it to the linker. 

An sample file (for the default field configuration) is here:-  https://github.com/jembi/JeMPI/blob/dev-jempi-754-linker-m_and_u_tally/JeMPI_Apps/JeMPI_Linker/src/test/resources/csv/basic.csv

Whilst it is in the linker it will accumtae the Field-Equal-Pair-Match matrix.  If you have enabled the debug level logging, you should see logs similar to this for the linker (i.e `docker service logs jempi_linker`)

```bash

empi_linker.1.u7lsr7bw710w@csymbionix    | [TRACE] 2023-12-20 03:42:21.586 BackEnd:155 - 2023/12/20 03:40:22:0000796
jempi_linker.1.u7lsr7bw710w@csymbionix    | [INFO ] 2023-12-20 03:42:21.636 FieldEqualityPairMatchProcessor:61 - FieldEqualityPairMatchProcessor: Processing 23 candidates
jempi_linker.1.u7lsr7bw710w@csymbionix    | [DEBUG] 2023-12-20 03:42:21.637 FieldEqualityPairMatchProcessor:75 - Saving candidates m and u values to kafka
jempi_linker.1.u7lsr7bw710w@csymbionix    | [DEBUG] 2023-12-20 03:42:21.637 FieldEqualityPairMatchProcessor:76 - ----- Field Data -----
jempi_linker.1.u7lsr7bw710w@csymbionix    | -> phoneNumber
jempi_linker.1.u7lsr7bw710w@csymbionix    | | ------------------------------------------|
jempi_linker.1.u7lsr7bw710w@csymbionix    | |               | Pair Match | Pair Unmatch |
jempi_linker.1.u7lsr7bw710w@csymbionix    | | ------------------------------------------|
jempi_linker.1.u7lsr7bw710w@csymbionix    | | Field Equal   |     0     |      0      |
jempi_linker.1.u7lsr7bw710w@csymbionix    | | Field Unequal |     0     |      23      |
jempi_linker.1.u7lsr7bw710w@csymbionix    | | ------------------------------------------|
jempi_linker.1.u7lsr7bw710w@csymbionix    | 
jempi_linker.1.u7lsr7bw710w@csymbionix    | 
jempi_linker.1.u7lsr7bw710w@csymbionix    | -> gender
jempi_linker.1.u7lsr7bw710w@csymbionix    | | ------------------------------------------|
jempi_linker.1.u7lsr7bw710w@csymbionix    | |               | Pair Match | Pair Unmatch |
jempi_linker.1.u7lsr7bw710w@csymbionix    | | ------------------------------------------|
jempi_linker.1.u7lsr7bw710w@csymbionix    | | Field Equal   |     0     |      0      |
jempi_linker.1.u7lsr7bw710w@csymbionix    | | Field Unequal |     0     |      23      |
jempi_linker.1.u7lsr7bw710w@csymbionix    | | ------------------------------------------|
jempi_linker.1.u7lsr7bw710w@csymbionix    | 
jempi_linker.1.u7lsr7bw710w@csymbionix    | 
jempi_linker.1.u7lsr7bw710w@csymbionix    | -> nationalId
jempi_linker.1.u7lsr7bw710w@csymbionix    | | ------------------------------------------|
jempi_linker.1.u7lsr7bw710w@csymbionix    | |               | Pair Match | Pair Unmatch |
jempi_linker.1.u7lsr7bw710w@csymbionix    | | ------------------------------------------|
jempi_linker.1.u7lsr7bw710w@csymbionix    | | Field Equal   |     0     |      1      |
jempi_linker.1.u7lsr7bw710w@csymbionix    | | Field Unequal |     0     |      22      |
jempi_linker.1.u7lsr7bw710w@csymbionix    | | ------------------------------------------|
jempi_linker.1.u7lsr7bw710w@csymbionix    | 
jempi_linker.1.u7lsr7bw710w@csymbionix    | 

```

Once it is done, you can the call the api url to get the current m and u for each field

**Important note:-** 

* Before running this you need to make should you have run the configurator, so as to create the appropiate field information needed whilst tallying the value. Also note:- if you are going to use the Async Reciever, the interaction csv's fields need to match those configured

* As it is right now all interaction go through the same linker without any ability to distinguish were particular interaction are coming from (i.e a different csv). Due to this limitation, the code will continue to tally the  Field-Equal-Pair-Match  as if it where from one source (regardless if the interaction data is from abother source)
